### PR TITLE
Initial build of the auth helper

### PR DESCRIPTION
**Pull request by muir**

### CODE REVIEW

**Comment by muir** [comment]
The github actions are currently pending approval/setup from SRE.   They may need to be replaced with gitlab pipelines.

**Comment by muir** [comment]
@hniemchenko : I built this passing things on the command line.  Does that work for you, or would you like to pass in data some other way?

@mitchwadair : I built in a default page.  Does that work for you or should it be something else?

File: authhelper.go
Comment (mitchwadair) [inline authhelper.go:30]: does this exist, or is that what I need to implement?  If we want it tied to portal auth, we'd want the `portal` subdomain instead: `https://portal.singlestore.com/engine-sso`

File: authhelper.go
Comment (mitchwadair) [inline authhelper.go:58]: since the browser session is going to be making a call to this local server, should we pick a static port number?  How will the browser know what port to make the call to?

File: authhelper.go
Comment (mitchwadair) [inline authhelper.go:132]: if I'm reading this right, all that needs to be passed as the body for this endpoint is the JWT string retrieved through the auth process right?

**Comment by mitchwadair** [comment]
> @mitchwadair : I built in a default page. Does that work for you or should it be something else?

I might have missed this, where exactly is it?

File: authhelper.go
Comment (muir) [reply authhelper.go:30]: Yeah, that's what you need to create.  I'll change the URL.

Comment (muir) [reply authhelper.go:30]: I pass the reply URL to the FE.
It's the `returnTo` parameter.  You must verify that it starts with "http://localhost:"

File: authhelper.go
Comment (muir) [reply authhelper.go:58]: @mitchwadair 
I pass that to FE.  It's the `returnTo` parameter.  You must validate that the URL starts with `http://localhost:`

File: authhelper.go
Comment (muir) [reply authhelper.go:135]: Yes.

**Comment by muir** [comment]
> 

That's that base

**Comment by muir** [comment]
> > @mitchwadair : I built in a default page. Does that work for you or should it be something else?
> 
> I might have missed this, where exactly is it?

That's the baseURL.

File: authhelper.go
Comment (toast2e) [inline authhelper.go:22]: Can this be given a more descriptive name?

File: authhelper.go
Comment (toast2e) [inline authhelper.go:24]: Based on how I'm seeing this used, 'Password' feels like a misnomer and its use could give some people pause/trigger security warnings. Does 'Token' make more sense or am I misreading this?

File: authhelper.go
Comment (toast2e) [inline authhelper.go:137]: Could use the const http.MethodPost rather than hard-coded string

**Review by toast2e — COMMENTED** [review]
I don't have much to add to the logic which looks ok to me. A few random nits about naming and constants but otherwise, LGTM. I'll let Mitch chime in with final approval.

File: authhelper.go
Comment (muir) [reply authhelper.go:137]: @toast2e : I don't think those constants aid readability and I've generally avoided them on purpose.

File: authhelper.go
Comment (muir) [reply authhelper.go:138]: Ditto.  I think `400` has more obvious meaning than `http.StatusBadRequest`.  The only advantage the constants have is a reduced chance of an un-caught typo.

File: authhelper.go
Comment (muir) [reply authhelper.go:22]: Changed to `AuthHelperOutput` and added a comment.  Good enough?

File: authhelper.go
Comment (muir) [reply authhelper.go:24]: Changed to `PasswordToken` and added a small comment.  Good enough?

**Review by toast2e — APPROVED** [review]
I'm good with it.

**Comment by mitchwadair** [comment]
lgtm

Is there a good way to have this open up a localhost page for local testing?  Worst case, it doesn't and I just manually type the URL for localhost with the `returnTo` param

**Review by mitchwadair — APPROVED** [review]

**Comment by muir** [comment]
> Is there a good way to have this open up a localhost page for local testing? Worst case, it doesn't and I just manually type the URL for localhost with the `returnTo` param

What's your test scenario?  You thinking of using this program in your test?

**Comment by mitchwadair** [comment]
> What's your test scenario? You thinking of using this program in your test?

Just for local development so I can test FE functionality as I go

**Comment by muir** [comment]
> > What's your test scenario? You thinking of using this program in your test?
> 
> Just for local development so I can test FE functionality as I go

Hmm, seems like what would be best would be for it to not exit when it receives a value?
That would be an easy flag to add.   What would help you the most?

**Comment by mitchwadair** [comment]
> Hmm, seems like what would be best would be for it to not exit when it receives a value? That would be an easy flag to add. What would help you the most?
Basically, instead of opening to `https://portal.singlestore.com/engine-sso` I'd need it to open the browser to `localhost:<PORT>/engine-sso`.  If this is too much, I'm sure I can just test by copying the query params from the URL it does actually open with to a localhost URL manually

**Comment by muir** [comment]
> Basically, instead of opening to `https://portal.singlestore.com/engine-sso` I'd need it to open the browser to `localhost:8001/engine-sso`. If this is too much, I'm sure I can just test by copying the query params from the URL it does actually open with to a localhost URL manually

I've added a `--baseURL=whatever` parameter.

File: authhelper.go
Comment (hniemchenko) [inline authhelper.go:39]: Are these and BaseURL for some future use? (these are not currently passed from the driver)

**Review by hniemchenko — APPROVED** [review]
Overall looks good to me, I'll start testing it with JDBC

File: authhelper.go
Comment (muir) [reply authhelper.go]: There shouldn't be.   I will remove the `omitempty`

File: authhelper.go
Comment (muir) [reply authhelper.go:39]: BaseURL is used for FE development.
ClusterID & Databases are for future use.

File: authhelper_test.go
Comment (muir) [reply authhelper_test.go]: Good catch.  Thanks.

**Pull request merged**
